### DISCUSSION TITLE
agesocial: Country-specific social contacts and demography

### DISCUSSION THREAD

**Discussion by adamkucharski** [epiverse-trace/epiverse-trace.github.io#303] [Project Ideas]
### Description

Epidemic scenario models often need country-specific demography and data on age-stratified social mixing, particularly if implemented as interactive dashboards where users select a country. In Epiverse package vignettes, contact matrices are currently pulled from available data in {[socialmixr](https://cran.r-project.org/web/packages/socialmixr/vignettes/socialmixr.html)}. Alternatively, synethic matrices can be generated with packages like [conmat](https://github.com/idem-lab/conmat).

However, other model implementations, such as [covidm shiny app](https://github.com/nicholasdavies/covidm_shiny2) and [simex](https://github.com/WHO-Collaboratory/simex) have backend functions that import demography and projected contact matrices for a range of countries from [Prem et al](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005697) or [contactdata](https://hugogruson.fr/contactdata/) that align with model structure (e.g. number of age classes), as well as pulling from alternative sources where available.

Given these are so commonly used, it could make sense to modularise this particular functionality for matrix-model alignment so that could be called by the various scenario models out there – either as a small package or a contribution to the above ones.

### Typical end-users

Users of epidemics, finalsize, simulist, odin, seir etc.

### Key potential collaborators

Maintainers of simex, socialmixr, conmat, contactdata

### Inputs

Social mixing data and demographic statistics

### Outputs

Age-specific social mixing matrix and demography vector

### Imports

simex and [covidm_shiny2](https://github.com/nicholasdavies/) include the underlying data files, but would probably not be best suited as dependencies – instead, they could use this modularised functionality.

### Used by

finalsize
epidemics
seir

### Related projects

* [socialmixr](https://cran.r-project.org/web/packages/socialmixr/vignettes/socialmixr.html)
* [conmat](https://idem-lab.github.io/conmat/dev/)
* [covidm_shiny2](https://github.com/nicholasdavies/covidm_shiny2)

### Usage

```bash
contact_data <- get_contacts(iso3="USA", age_groups=c(0,20,40) )

demography_vector <- get_demography(iso3="USA", age_groups=c(0,20,40) )
```

### Additional comments

...

**Comment by adamkucharski** [comment]
Would be interested to hear thoughts from @sbfnk @finlaycampbell @nicholasdavies @kieshaprem @njtierney
and others, as some of this functionality may well already be available somewhere I've missed, or there may be better ways to solve this problem that the above proposal!

**Comment by Bisaloo** [comment]
Matrices from Prem et al. are packaged in [contactdata](https://github.com/Bisaloo/contactdata). Is any functionality you need not available there?

**Reply by adamkucharski** [reply]
Thanks. Have edited above to try and make clearer. Basically, was reviewing how contacts handled in [covidm shiny dashboard](https://github.com/nicholasdavies/covidm_shiny2/blob/master/makecontact_shiny2.R) and [simex](https://github.com/WHO-Collaboratory/simex/blob/main/R/add_cdat.R) and looks like there are two additional components that (I don't think) are currently easily available for users:
- Providing 'best available' contact data for a given country selection (e.g. from a survey if available, or synthetic/projected data if not). This isn't a particular issue for a single country run, but can be a challenge for multi-country dashboards.
- Aggregating age bands to align with model structure, scaling contact rates appropriately.

E.g. is there functionality that could make this line in {epidemics} README run with any country?

```r
contact_data <- socialmixr::contact_matrix(
  polymod,
  countries = "United Kingdom",
  age.limits = c(0, 20, 40),
  symmetric = TRUE
)
```

Know there is already some ongoing effort to align matrix classes etc across packages, so this may well already be planned.

**Reply by finlaycampbell** [reply]
I wasn't aware of contactdata, I would've just called that if I'd known about it! Options for (dis)aggregating the age bands would be v useful. We don't really have any big long-term plans for simex but happy to make adjustments to improve interoperability with other packages.

**Reply by Bisaloo** [reply]
> Aggregating age bands to align with model structure, scaling contact rates appropriately.

Yes, the aim for this is that it should be part of the shared S3 class many people listed in the first message are working towards: https://socialcontactdata.github.io/contactmatrix/articles/design-principles.html

**Reply by adamkucharski** [reply]
Sounds great. Tagging @avallecam as we've been discussing having more on social contacts in future iterations of training materials.

**Reply by adamkucharski** [reply]
Also tagging @zmcucunuba for info

**Comment by sbfnk** [comment]
For demographics `socialmixr` uses the `wpp2017` package (there are different ones for different releases of the World Population Prospectus), see e.g. the [`wpp_age()`](https://epiforecasts.io/socialmixr/reference/wpp_age.html) and [`pop_age()`](https://epiforecasts.io/socialmixr/reference/pop_age.html) functions -- although as we're working towards [making things work less magically](https://github.com/epiforecasts/socialmixr/issues/161#issuecomment-2501235901) in `{socialmixr}` this might actually be removed in favour of showing people how they can do this themselves using `wpp2017` (or a better source if available).

I'm not convinced there is such a thing as a "best available" contact matrix (the modelled ones may well be more suited to some use cases). Once we streamline the returned S3 classes people will be able to pick & choose between modelled and observed contact data using the two packages which hopefully will provide a better user experience.

What I would say is that it would be really useful to have a package that refits the Prem et al. model with arbitrary age groups, as both disaggregating and aggregating these would be pretty hard, and it would be good to have a way to reflect uncertainty in the estimates.
